Tools/Scripts/libraries/webkitflaskpy/webkitflaskpy/ipc_manager.py

```diff
@@ -0,0 +1,122 @@
+# DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+# SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+# CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+# OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+# OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+import hashlib
+import json
+
+from webkitcorepy import string_utils
+from webkitflaskpy import Database
+
+
+class IPCManager(object):
+     DEFAULT_EXPIRATION = 10 * 60
+     DEFAULT_TIMEOUT = 60
+     RETRY = 3
+
+     def __init__(self, *databases):
+         self.databases = databases or [Database()]
+         self.callbacks = dict()
+
+     def register(self, name, callback):
+         if not isinstance(name, str):
+             raise ValueError('Name of registered callback must be a string')
+         if not callable(callback):
+             raise ValueError('Callback must be callable')
+         self.callbacks[name] = callback
+
+     def enqueue(self, name, *args, **kwargs):
+         expiration = kwargs.pop('ex', self.DEFAULT_EXPIRATION)
+         retry = kwargs.pop('retry', self.RETRY)
+         if name not in self.callbacks:
+             raise ValueError("'{}' is not a registered callback".format(name))
+         encoded = json.dumps([name, args, kwargs, retry, expiration])
+         digest = hashlib.md5()
+         digest.update(string_utils.encode(encoded))
+         digest = digest.hexdigest()
+
+         for database in self.databases:
+             database.set(digest, encoded, ex=expiration)
+         return True
+
+     def dequeue(self, key, filter=None, index=None, timeout=None):
+         timeout = timeout or self.DEFAULT_TIMEOUT
+         if index is None:
+             for i in range(len(self.databases)):
+                 result = self.dequeue(key, index=i, timeout=timeout)
+                 if result:
+                     return result
+             return False
+
+         lock = self.databases[index].lock(name='lock_{}'.format(key), timeout=timeout, blocking_timeout=0.05)
+         try:
+             if not lock.acquire():
+                 return False
+             raw = self.databases[index].get(key)
+             if not raw:
+                 self.databases[index].delete(key)
+                 return False
+             data = json.loads(raw)
+             if filter and not filter(data):
+                 return False
+             if len(data) != 5:
+                 self.databases[index].delete(key)
+                 return False
+             name, args, kwargs, retry, expiration = data
+             if name not in self.callbacks:
+                 self.databases[index].delete(key)
+                 return False
+
+             if retry > 0:
+                 self.databases[index].set(key, json.dumps([name, args, kwargs, retry - 1, expiration]), ex=expiration)
```

Comment:
Ditto, we may want to `sort_keys=True`. Since we keep repeating  json.dumps in this file and unit test, I guess it would be more reasonable to have a helper function called `encode_task` which produces encoded task. It would also be nice to have `task_key` function, which centralizes the key calculation and can be used in unit test as well.